[RNMobile] Enable Third-Party Blocks to Set Aspect Ratio of Image via Inner Blocks

SiobhyB opened the pull request:
## Description

The React Native version of [Jetpack's Tiled Gallery block](https://jetpack.com/support/tiled-galleries/) will be using [the Inner Block API](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/inner-blocks) to display images in a gallery. As part of this project, we need to change the aspect ratio of the images that are displayed.

It is likely that other third-party blocks will face similar use cases in the future if they're pulling in the image block as an inner block. With that in mind, and to help achieve our end goal, this PR introduces some changes that will enable any third-party block to define the aspect ratio of an image when used as an inner block.

_Note, the goal behind this PR _isn't_ to introduce Jetpack-specific code, but rather to introduce generic changes that will hopefully be useful or re-usable for other third-party blocks. We have tried to be mindful against introducing anything that is specific only to Jetpack to the Gutenberg codebase._

## Types of changes

Outlined below are specifics around the two main changes that this PR includes. They are the introduction of a new context, `fixedHeight`, to prevent an overlap with the core Gallery block's set height and a tweak to the image block's `shapeStyle` prop.

### Address Conflict with Core Gallery Block's Style 

* The [core Gallery block](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/gallery) introduced [two block contexts](https://github.com/WordPress/gutenberg/blob/bee52e68292357011a799f067ad47aa1c1d710e1/docs/reference-guides/block-api/block-context.md) in order to get around the challenge of editing/styling the image block as an inner block. Those are `allowResize` and `imageCrop`. 
* The `imageCrop` context [translates to `resizeMode='cover'`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js#L633), which is what we need to use in order to change the inner image's aspect ratios.
* The use of `imageCrop`, meant [the image block's `hasImageContext` variable](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js#L517) began to evaluate as `true` for the Tiled Gallery block. This then meant [the `fixedHeight` style](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js#L636) was being applied to it, which prevented the aspect ratio of images from being changed. 
* With the above in mind, this PR introduces a [new `fixedHeight` context](https://github.com/WordPress/gutenberg/pull/35089/files#diff-aa7f347bbb8d1aa1c43808397ef62ef1d0b789773b127768c01c913d1a956e19R96) to the Gallery block. As explained [here](https://github.com/WordPress/gutenberg/pull/35089/files#r717078745), this change prevents styles that are specific to the core Gallery block from impacting any block that uses the `imageCrop` context.

### Allow Third-Party Styles to be Passed

* The image block's [`shapeStyle` prop was updated](https://github.com/WordPress/gutenberg/pull/35089/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R698) to accept not only a string, but also full styles. This is to allow third-party blocks, the Tiled Gallery block in this case, to pass styles without needing to make edits directly to the image block's core files.

## How has this been tested?

As this PR is part of a wider effort to port the Jetpack Tiled Gallery block to Gutenberg Mobile, steps to specifically test the Jetpack block can be found in the Jetpack repository at https://github.com/Automattic/jetpack/pull/21166, with the central Gutenberg Mobile PR at https://github.com/wordpress-mobile/gutenberg-mobile/pull/4017.

For this specific PR, on the Gutenberg side, the following has been tested in the iOS and Android apps, with this branch checked out from view:

* Ensure the images in the Gallery block are a set height and that there are no regressions.
* Add an Image block and ensure the image displays at the correct size. Play around with the block's various settings, including the round style, and ensure there are no regressions. 

## Screenshots

These screenshots show our specific use cases for using contexts, which is to remove and style parts of the image block when used as an inner block. You'll see we were able to use contexts to make the images appear as squares:

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/134989742-1b86fb15-1699-4b21-ae22-612527151a21.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/134991240-cfeac12b-4dd7-4466-b3ca-55948d09149a.png" width="100%"> |

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

SiobhyB committed:
Add 'is-style-squared' CSS selector

This selector will be used to make images in the editor look 'square' within the Tiled Gallery block.

The CSS will be iterated on in future commits.

SiobhyB committed:
Hide block's caption component in Tiled Gallery

As the image block's caption component isn't part of the Tiled Gallery's design, this commit sets up the necessary code to hide it from view.

This is done through the use of a "hideImageCaption" context, which lives under a new "blockStyles" object. It should be possible to further utilise this object for future styling of inner blocks.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/add/squared-aspect-ratio

SiobhyB committed:
Clarify purpose of new object

With this new commit, the purpose of the new object is clarifyied by tweaking its name from "blockStyles" to "imageBlockStyles".

SiobhyB committed:
Set the image's aspect ratio, so that it's a square

SiobhyB committed:
Merge branch 'rnmobile/add/squared-aspect-ratio' of https://github.com/WordPress/gutenberg into rnmobile/add/squared-aspect-ratio

SiobhyB committed:
Only used fixed height if it's explicitly set

At the moment, a fixed height is applied to the image block if context is detected. This not only effects the standard Gallery block (intended) but also the Tiled Gallery Block (not intended).

With this commit, a new "fixedHeight" context is added to only set a fixed height if it's explicitly set.

This is necessary as the Tiled Gallery block will need varying heights as part of its design.

SiobhyB committed:
Allow third-party blocks to set "shapeStyle"

In order to change the image's shape via a third-party block (such as Jetpack's Tiled Gallery), we need to be able to pass a style to the "shapeStyle" prop.

With this commit, the "shapeStyle" prop is changed to accept both a reference to the image block's classes and a third-party block's styles.

This change means that it's no longer necessary for third-party blocks, like the Tiled Gallery, to set styles directly in the image block's stylesheet. So, this commit also removes the ".is-style-squared" class.

SiobhyB committed:
Enable context objs to style images in blocks

The 'hideImageCaption' and 'fixedHeight' context objects are enabled for the image block in this commit. These objects will be used for the purpose of tweaking the image block's default styles in blocks where it's used i.e. currently, the Gallery block and Jetpack's Tiled Gallery block.

SiobhyB committed:
Improve conditional code by removing redundancy

SiobhyB commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -631,9 +631,13 @@ export class ImageEdit extends Component {
 		};
 
-		const imageContainerStyles = [ hasImageContext && styles.fixedHeight ];
+		const imageContainerStyles = [
```
👋  @geriux, I wanted to check in with you, as this PR makes some changes to the ones you introduced with your recent PR in https://github.com/WordPress/gutenberg/pull/34957. 

The reason for the changes is that `hasImageContext` was evaluating to `true` for the Tiled Gallery block, and the block needs to have more flexible heights. You'll see from this PR that I've worked around this by adding a new `fixedHeight` context object to the Gallery block, so it's only used with that block. 

Does that seem like an okay approach to you? Interested to know whether you have other thoughts.

geriux replied:
Since the only block that was using context was the Gallery block I used a generic `fixedHeight` name to avoid limiting the height of the component.

Seeing that now the Tiled block would need this as well, what if we use a name like `isGallery` instead of `fixedHeight` for the context?

Do we also need to pass `fixedHeight` from the Image block?

SiobhyB replied:
@geriux, thank you for looking into this! 🙇‍♀️ 

> Seeing that now the Tiled block would need this as well, what if we use a name like `isGallery` instead of `fixedHeight` for the context?

Ah, the Tiled Gallery block doesn't need this context. I created it as `hasImageContext` was evaluating as true for the Tiled Gallery block and applying the fixed height that was needed only for the core Gallery block. This was preventing images from being styled as squares within the Tiled Gallery block. That's what led me to create a context that would only apply to the core Gallery block, not other blocks that are using contexts.

I'm happy to rename it if you want it be clearer that this is for use with the core gallery, maybe something like `galleryHasFixedHeight` or `isCoreGallery`?

> Do we also need to pass `fixedHeight` from the Image block?

That's a good point! It doesn't really need to be passed as an `additionalImageProp`. Do you think a new `imageBlockStyles` object would make sense here?

SiobhyB replied:
I've gone ahead to create an `imageBlockStyles` object and moved the `fixedHeight` context to it in https://github.com/WordPress/gutenberg/pull/35089/commits/ce9ec06e1c0e5b3a75a2a985f2f127a472d482ee.

SiobhyB committed:
Merge branch 'rnmobile/add/tiled-gallery-block' into rnmobile/add/squared-aspect-ratio

SiobhyB committed:
Remove 'hideImageCaption' context to limit PR's scope

In order to limit this PR's scope, this commit removes the 'hideImageCaption' context. Instead, the focus will be on changes needed for a third-party block to change the image's aspect ratio. The proposal for contexts to hide certain settings in the image block will be moved to https://github.com/WordPress/gutenberg/pull/35315.

SiobhyB committed:
Stop uneccessarily passing context as an image prop

As mentioned in https://github.com/WordPress/gutenberg/pull/35089#discussion_r720023932, it's not necessary to pass the 'fixedHeight' as an 'additionalImageProps'. This commit moves that context to a new 'imageBlockStyles' object. This object has the potential to be added to in the future.

SiobhyB commented:
Noting that the original scope of this PR has been refined. 

The PR originally included the introduction of a new context, `hideImageCaption`, to the image block. As the introduction of new contexts that are specifically designed with third-party blocks in mind is new and will require further discussion, that aspect is going to be moved to a separate PR.

Now, this PR's scope is limited to changes that are necessary for third-party blocks to change the aspect ratio of images pulled in via inner blocks. It involves a tweak to the image block's `shapeStyle` prop and a change that prevents a conflict with the core Gallery block's styles.

SiobhyB commented:
I'm closing this PR as, to make things as clear and easy to follow along with as possible, it's necessary to split it up even further into the following:

1. A PR to tweak the image block's `shapeStyle` prop, this can be found at https://github.com/WordPress/gutenberg/pull/35420.
2. A PR to address the conflict with the core Gallery block's styles. This is a WIP at the time of writing.
